feat: add local langfuse tracing option

ahau-square opened the pull request:
**why**
The purpose of this PR is to integrate local [Langfuse](https://langfuse.com/) tracing into the project to enhance debugging and monitoring capabilities. Tracing allows developers to observe the flow of execution and diagnose issues more effectively.

**what**

_Exchange Package:_
     - Defined `observe` decorator wrapper `observe_wrapper` to use the observe decarator only if Langfuse local env variables are set
     - Add observe decorator to tool calling function and to providers' completion functions

_Goose:_
     - Modifications in  to set up Langfuse tracing upon CLI initialization.
     - Updates in  to trace session-level information.
     - Add observe decorator to reply

**usage**
- Developers can use [locally hosted Langfuse](https://langfuse.com/docs/deployment/self-host) tracing by applying the `observe_wrapper` decorator to functions for automatic integration with Langfuse, providing detailed execution insights. Docker is a requirement.
- To enable tracing, launch the CLI with the `--tracing` flag, which will initialize Langfuse with the set env variables

**setting up locally hosted Langfuse**
- Run `setup_langfuse.sh` script (in `packages/exchange/src/exchange/langfuse`) to download and deploy the Langfuse docker container with default initialization variables found in the `.env.langfuse.local` file. 
- Go to http://localhost:3000/ and log in with the default email/password output by the shell script (or found in the  `.env.langfuse.local` file).
- Run Goose with the --tracing flag enabled i.e., `goose session start --tracing`
- When Goose starts you should see the log INFO:exchange.langfuse:Langfuse credentials found. Find traces, if enabled, under http://localhost:3000.
- Go to http://localhost:3000/ and you should be able to see your traces


**Sample trace viewing:**

<img width="1153" alt="Screenshot 2024-10-02 at 4 45 50 PM" src="https://github.com/user-attachments/assets/0f521b95-80d1-4ff9-90fc-85253a34663c">

lifeizhou-ap commented:
Hey @ahau-square, 

The Langfuse UI looks good for users to view the traces! Below is what I found about tracing with Langfuse

**Pros**
- UI is clear easy to browse
- Api is easy to use and clean
- build-in LLM specific tracing 

**Cons**
- Manual steps to setup the Langfuse account
We could copy the docker compose file in our repo to start the docker container instead of asking the user to clone the langfuse repo.   (The drawback is that the docker compose file could be out of date. However, in the docker compose file there are not much customised logic, it mainly defined the image and basic configurations
However, I found it is hard to automate the steps to sign up as a user, create a project and get the api keys

**Alternative**
Open telemetry can integrated with a few tools that visualise the tracing.   For example, export to local Zipkin or using otel-tui as @codefromthecrypt suggested. These tools can be started as docker container via scripts without manual steps.

For your reference,  this is the [PR](https://github.com/block-open-source/goose/pull/75) @codefromthecrypt created for tracing.

lifeizhou-ap commented:
When the user passes --tracing, but langfuse server is not up.  it shows error "ERROR:langfuse:Unexpected error occurred. Please check your request and contact support: https://langfuse.com/support."

Maybe we can validate whether the server is up when the user pass `--tracing` option and remind the users to start the server

codefromthecrypt commented:
ironically just got off a call with @marcklingen. So marc, "goose" as you can read in its lovely readme is an inference backed SWE capable buddy. It uses its own LLM abstraction called exchange. I've thought about instrumenting this to use the genai conventions as tagged above. I actually haven't gotten around to spiking that, either.

My thinking is that we can maybe do both (langfuse and otel) until langfuse handles OTLP? Then compare until the latter works as well. wdyt? Do you have an issue to follow-up on that?

Also, if you can help @ahau-square meanwhile with any pointers or any notes about comments above. We've been using python VCR tests at times, so we can probably test things work even without accounts. Depending on how things elaborate I don't mind contributing help on that as it is a good way to learn.

Meanwhile, I think let's see where we get and keep a debug story going, even if it is work in progress

marcklingen commented:
thanks for tagging me @codefromthecrypt, happy to help, "goose" seems cool and I'll try to give it a proper spin later this week

> My thinking is that we can maybe do both (langfuse and otel) until langfuse handles OTLP? Then compare until the latter works as well. wdyt? Do you have an issue to follow-up on that?

This seems very reasonable

I am generally very excited about standardization on the instrumentation side via opentelemetry. We are tracking support for OTLP in Langfuse here, feel free to subscribe to the discussion or add your thoughts: https://github.com/orgs/langfuse/discussions/2509

> However, I found it is hard to automate the steps to sign up as a user, create a project and get the api keys. It is a once-off step for the user. If we could automate it, the user will have smoother experiences

> Also, if you can help @ahau-square meanwhile with any pointers or any notes about comments above. We've been using python VCR tests at times, so we can probably test things work even without accounts. Depending on how things elaborate I don't mind contributing help on that as it is a good way to learn.

Langfuse supports [initialization via environment variables](https://langfuse.com/changelog/2024-09-04-headless-initialization-of-self-hosted-deployments) which would overcome the issue of having to create an account via the local langfuse server. I will add this to the default docker compose configuration as well and update this message once done.

Side note: I think interesting here would be to return a [deep link to the trace](https://langfuse.com/docs/tracing-features/url) in the local langfuse ui on the cli output if tracing is enabled to make it easier to jump to the trace from any invocation of goose for the debugging use case.

marcklingen commented:
Added support for headless initialization via Docker Compose here: https://github.com/langfuse/langfuse/pull/3568/files.

This change required explicitly handling empty string values for these environment variables, as Docker Compose does not allow adding environment variables that may not exist without setting a default value. Therefore, this depends on the next langfuse release which will include this fix (scheduled for tomorrow).

marcklingen commented:
we just [released](https://github.com/langfuse/langfuse/releases/tag/v2.82.0) the change. let me know if you have any questions, @ahau-square. You can now pass the init envs to the Langfuse docker compose deployment to initialize default credentials.

marcklingen commented:
There's no way to bypass it. They'd need to sign in with the default credentials to view the data in the UI.

marcklingen commented:
addition to the above @ahau-square, if this is local anyway, you could make all traces public. thereby users could directly jump from a trace url in their console output to viewing the trace in their local langfuse instance without having to sign in. this will limit the features that are available though.

more on this here: https://langfuse.com/docs/tracing-features/url#share-trace-via-url

codefromthecrypt commented:
nit: is this a chore or a feature? the PR title shows chore, but I suspect the result is more a feature than say, reformat files. wdyt?

lifeizhou-ap commented:
Nice one! @ahau-square 

A couple of suggestions:

- Instead of checking out the langfuse repository, how about in the script we copy `docker-compose.yaml` file from the latest langfuse gitrepo and then start the containe? In this case, user has one step less to setup the langfuse local server and we don't need to have Langfuse code in our local goose project.  

- The `langfuse` folder has files with mixed purposes.
    - `langfuse.py`: used in the `goose` (with exchange) application.  
    - `setup_langfuse.sh`: used to start Langfuse server
    - `env.langfuse.local`:  used for both `goose` and Langfuse server
    
Since Langfuse server is a self contained application, we could create a fold outside of src/exchange (could be under `goose` such as `Langfuse-server`, It will contain `setup_langfuse.sh`, env file containing the variables for starting Langfuse, and docker-compose file

lifeizhou-ap commented:
@ahau-square was asking what kinds of tests should be written for this PR. 

In terms of testing, I feel like we can make it simple just to write unit tests

-  observe_wrapper function. Test cases
    - langfuse_context observe function is called if the credential env variable exists apart from executing the function
    - if the credential does not exist, it executes function only
- maybe we can also write tests on openai.py…. to make sure the decorator is applied when credential is present. 

I found it is a bit hard to test at the integration level.  Although we can run the tests and start the Langfuse server, we still have to manually login to verify the traced data unless there is existing Langfuse api for us to verify

Another alternative is to test the integration with Langfuse, but it requires us to know the implementation details of the langfuse_context.observe.  I feel it is a bit overkill for us to test in this way too.    FYI These looks like the relevant [implementation](https://github.com/langfuse/langfuse-python/blob/7adec675a51f84747ff9c6dd389a4881276d111e/langfuse/decorators/langfuse_decorator.py#L102) and the [tests](https://github.com/langfuse/langfuse-python/blob/7adec675a51f84747ff9c6dd389a4881276d111e/tests/test_decorators.py) in Langfuse python. 

It would be great if you could give us some advice @marcklingen  :) Thank you!

marcklingen commented:
Usually testing this is overkill for most teams as you mentioned.

If you however want to test it, I've seen some teams run an example and then fetch the same trace id (fetch_trace) to check that it includes all of the observations that it should include. This however necessitates that you (1) run langfuse in ci via docker compose, (2) use flush to make sure the events are sent immediately at the end of the test, and (3) wait e.g. 2-5sec in CI as Langfuse does not have read after write consistency on the apis.

lifeizhou-ap commented:
> Usually testing this is overkill for most teams as you mentioned.
> 
> If you however want to test it, I've seen some teams run an example and then fetch the same trace id (fetch_trace) to check that it includes all of the observations that it should include. This however necessitates that you (1) run langfuse in ci via docker compose, (2) use flush to make sure the events are sent immediately at the end of the test, and (3) wait e.g. 2-5sec in CI as Langfuse does not have read after write consistency on the apis.

Thanks for that @marcklingen!  I feel that we could skip this kind of integration test until tracing becomes a critical feature for our application, but it is good to know about how people test!

codefromthecrypt commented:
Even if we don't run tests in CI right now, probably worth describing them or even having a mechanism with default config in the repo, mentioned in CONTRIBUTING.md? Usually fixtures like this are quite tricky to setup the first time. If we punt them, then someone potentially not on this issue yet may have a large task which is to figure out what's wrong, how to test it locally, etc while under pressure. "An ounce of prevention is worth a pound of cure." sort of thing

marcklingen commented:
Seems like this is roughly an hour to set up. You can have a look at how we run Langfuse via docker compose in the Langfuse/Langfuse main CI pipeline (public GH Actions)

codefromthecrypt commented:
@marcklingen thanks. Here's a suggestion forward: if we don't add an integrated test or an integrated ad-hoc capability in this PR, add something to CONTRIBUTING.md like this.

Langfuse is experimental, and we don't currently have integration tests. To try this locally with docker, use steps similar to [Langfuse CI pipeline](https://github.com/langfuse/langfuse/blob/13655515a216a0abff56fc0f1e8d3f69ee7a18d9/.github/workflows/pipeline.yml#L244).

We could still investigate [VCR tests](https://pytest-vcr.readthedocs.io/en/latest/) to have some sort of test working. For example, record requests made to a saas or ad-hoc install of langfuse. If decide to do no tests, maybe someone should sign up to manually verify before merge?

hope the suggestions help!

ahau-square commented:
Thanks all for the feedback! I've moved the langfuse-related code into a new package and also updated to only pull the docker compose file from the langfuse repo.

I'm going to skip the integration tests for now as this is still experimental. If it gets more traction we can revisit and left a note in the README in the langfuse-wrapper package to that effect. @codefromthecrypt if you want to take a stab at something more please feel free (tbh I'm not totally sure I understand what you're suggesting). I've implemented a couple tests per @lifeizhou-ap 's guidance.

lifeizhou-ap commented:
Hi @ahau-square The PR looks good! 

Just wondering why we create a project for `langfuse_wrapper` instead of leaving it in `exchange`? I guess it might be reused in other projects in addition to `exchange` or `goose` and would be published it as a package?  If it is not going be reused, we might not need to have a separate project for `langfuse_wrapper` 

I think both way works. I am just curious about it. :)

lamchau commented on .github/workflows/ci.yaml:
```diff
@@ -48,6 +48,27 @@ jobs:
+      - uses: actions/checkout@v4
+
+      - name: Install UV
+        run: curl -LsSf https://astral.sh/uv/install.sh | sh
```
:nit: we've been switching `uses: astral-sh/setup-uv@v3` this way we lock versions

ahau-square replied:
going to just wait on this one until they're all updated here

lamchau commented on packages/langfuse-wrapper/scripts/setup_langfuse.sh:
```diff
@@ -0,0 +1,70 @@
+    docker compose --env-file "$LANGFUSE_ENV_FILE" -f "$LANGFUSE_DOCKER_COMPOSE_FILE" up --detach
+}
+
+wait_for_service() {
```
we might want to add a max retries guard since this has the potential for looping indefinitely. probably an edge case but nice little ux in the event it does!

:nit: `nc` gives minimal output, `curl` is a little handier esp since we're using the http context 

```sh
    local retries=30
    local count=0
    until curl -s http://localhost:3000 > /dev/null; do
        ((count++))
```

lamchau commented on packages/langfuse-wrapper/scripts/setup_langfuse.sh:
```diff
@@ -0,0 +1,70 @@
+}
+
+start_docker_compose() {
+    docker compose --env-file "$LANGFUSE_ENV_FILE" -f "$LANGFUSE_DOCKER_COMPOSE_FILE" up --detach
```
:nit: what do you think if add a dependency check for commands so we can guard as early as possible? 

i like the pattern of checking all dependencies in a loop and print out the missing dependencies so i can copy and paste to install all of them at once (e.g. `apt install -y a b c`).

```sh
check_dependencies() {
    local dependencies=("curl" "docker" "docker-compose")
    local missing=()
}
```

lamchau commented on packages/langfuse-wrapper/scripts/setup_langfuse.sh:
```diff
@@ -0,0 +1,70 @@
+
+set -e
+
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"
```
:nit: `SCRIPT_DIR=$(realpath "$(dirname "${BASH_SOURCE[0]}")")` is a little more straightforward since it avoids invoking an extra subshell. 

if we want to stick with `cd`, could we use `pushd`/`popd` which makes pathing history/traversal easier in the event we need to debug and can get a full pathing history via `dirs`?

lamchau commented:
this is awesome! pretty excited to try this out :)

lamchau commented on packages/langfuse-wrapper/src/langfuse_wrapper/langfuse_wrapper.py:
```diff
@@ -0,0 +1,80 @@
+    logger.addHandler(handler)
+
+CURRENT_DIR = os.path.dirname(os.path.realpath(__file__))
+PACKAGE_ROOT = os.path.dirname(os.path.dirname(os.path.dirname(CURRENT_DIR)))
```
this works but i feel we could make this a bit more dynamic using `pathlib` and some marker files

something like this
```python
def find_package_root(start_path: Path, marker_file=('pyproject.toml')) -> Path:
```

if that seems too extra then maybe we can path up explicitly
```python
current_file = Path(__file__).resolve()
package_root = current_file.parents[3]
```

ahau-square commented:
> I think both way works. I am just curious about it. :)

We need the wrapper in the block plugins repo also since we want to wrap the block provider completions there as well.

lamchau commented on packages/langfuse-wrapper/scripts/setup_langfuse.sh:
```diff
@@ -15,21 +15,41 @@
+
+    if [ ${#missing_dependencies[@]} -ne 0 ]; then
+        echo "Missing dependencies: ${missing_dependencies[*]}"
+        echo "You can install them with: apt install -y ${missing_dependencies[*]}"
```
ohh sorry this is on me - i wasn't clear just `${missing_dependencies[*]}` would be enough! we don't know what package managers people have

lamchau reviewed (approved):
lgtm!

baxen commented on packages/langfuse-wrapper/src/langfuse_wrapper/langfuse_wrapper.py:
```diff
@@ -0,0 +1,96 @@
+    HAS_LANGFUSE_CREDENTIALS = True
+    logger.info("Langfuse context and credentials found.")
+else:
+    logger.warning(
```
This warning will display on any new goose session, which i think folks will find confusing! especially if they haven't heard of this integration. I'd remove this i think

baxen commented on packages/langfuse-wrapper/src/langfuse_wrapper/langfuse_wrapper.py:
```diff
@@ -0,0 +1,96 @@
+from functools import wraps  # Add this import
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
```
this will change the logging behavior of any main processes that import this file, which we want to avoid. I'd leave the log level and formatting configuration to the main process

baxen commented on packages/langfuse-wrapper/src/langfuse_wrapper/langfuse_wrapper.py:
```diff
@@ -0,0 +1,96 @@
+PACKAGE_ROOT = find_package_root(CURRENT_DIR)
+
+LANGFUSE_ENV_FILE = os.path.join(PACKAGE_ROOT, "env", ".env.langfuse.local")
+print(LANGFUSE_ENV_FILE)
```
nit: rm

baxen commented on src/goose/cli/session.py:
```diff
@@ -70,10 +73,9 @@ def __init__(
         self.status_indicator = Status("", spinner="dots")
         self.notifier = SessionNotifier(self.status_indicator)
-
+        langfuse_context.configure(enabled=tracing)
```
this line outputs:
```
Langfuse client is disabled. No observability data will be sent.
```

when langfuse is not setup (e.g. no --tracing) - should probably call this only if tracing is True?

marcklingen replied:
it might happen that this line is logged once when the sdk is loaded but tracing is disabled.

@hassiebp did we already or will we change this behavior?

ahau-square replied:
The observe_wrapper is imported in all the exchange providers with a completion function before `langfuse_context.configure(enabled=tracing)` is even called - so langfuse is actually initialized before the tracing flag is parsed if there is an available langfuse context (ie you are running local langfuse). So configuring the context to be disabled is required if that's the case to disable the tracing.

ahau-square replied:
I've updated it so that the log level is set to error and the message is suppressed in case --tracing is not set. Also updated some prints for if tracing is enabled but langfuse is not available.

@baxen  should we exit if --tracing was passed but langfuse isn't available on start up? Or just allow the session to continue running?

ahau-square replied:
decided to exit if --tracing is passed but langfuse not available

baxen reviewed (approved):
✨ Looks great! working well for me locally

Left some comments about verbosity of outputs and some log handling

marcklingen commented:
Nice @ahau-square, does it make sense to add a quick get-started to the docs? Happy to help
